RavenDB-18166 If query results come from different index defintions on shards then we mark the results as stale

arekpalinski opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18166/Different-index-defintions-after-a-query

### Additional description

Storing raft index on any modification of an index definition so we're able to detect on the orchestrator that query results returned by indexes having different definitions. In that case we mark the results as stale.

### Type of change

- Sharding feature

### How risky is the change?

- Moderate 

### Backward compatibility

- Non breaking change

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing by Contributor

- Tests have been added that prove the fix is effective or that the feature works
 - Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 

### Testing by RavenDB QA team

- No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

arekpalinski committed:
RavenDB-18166 If query results come from different index defintions on shards then we mark the results as stale

ppekrol commented on src/Raven.Server/Documents/Indexes/Index.cs:
```diff
@@ -3047,7 +3047,10 @@ public IndexDefinition GetIndexDefinition()
             IndexQueryServerSide query, QueryOperationContext queryContext, OperationCancelToken token)
         {
-            var result = new StreamDocumentQueryResult(response, writer, token);
+            var result = new StreamDocumentQueryResult(response, writer, token)
```
move this index to ctor

arekpalinski replied:
done

ppekrol commented on src/Raven.Client/Documents/Indexes/IndexUpdateClusterState.cs:
```diff
@@ -1,19 +1,22 @@
 ﻿namespace Raven.Client.Documents.Indexes
 {
-    internal class ClusterState
+    public class IndexUpdateClusterState
```
can we stay with internal?
can we rename to IndexDefinitionClusterState?

arekpalinski replied:
done

karmeli87 commented on src/Raven.Server/Json/BlittableJsonTextWriterExtensions.cs:
```diff
@@ -689,11 +689,11 @@ public static async Task<(long NumberOfResults, long TotalDocumentsSizeInBytes)>
                 writer.WriteComma();
-                writer.WritePropertyName(nameof(result.RaftCommandIndex));
-                writer.WriteInteger(result.RaftCommandIndex.Value);
+                writer.WritePropertyName(nameof(result.IndexDefinitionRaftIndex));
```
Is it a breaking change?

arekpalinski replied:
No, it was introduced in 6.0:

https://github.com/ravendb/ravendb/pull/14112/files#diff-34342d101561a370130f2ace8940f2a0f876375b0e9f5e693a31c98a76f7a0edR62-R63

arekpalinski committed:
RavenDB-18166 Addressing PR comments:
- IndexUpdateClusterState renamed to IndexDefinitionClusterState and it stayed with the internal access modifier
- introducing additional constructor parameter to query result classes (long? indexDefinitionRaftIndex)

ppekrol commented on src/Raven.Server/Documents/Queries/Dynamic/DynamicQueryRunner.cs:
```diff
@@ -62,7 +62,6 @@ public override async Task<DocumentQueryResult> ExecuteQuery(IndexQueryServerSid
             using (QueryRunner.MarkQueryAsRunning(index.Name, query, token))
             {
                 var queryResult = await index.Query(query, queryContext, token);
-                queryResult.RaftCommandIndex = result.Index;
```
Now we are always adding this, right? How does it affect Orchestrator, IIRC there was a piece of code that was waiting for that command to be applied, but it was only executed when this was null.

arekpalinski replied:
1. Yes, we always add it now.
2. We wait only if it's an map-reduce index

https://github.com/ravendb/ravendb/blob/af779cea08874f70f7867156abf57c2095b6783c/src/Raven.Server/Documents/Sharding/Queries/AbstractShardedQueryProcessor.cs#L454-L456

arekpalinski replied:
Indeed that was problematic because it caused that we called `WaitForExecutionOnAllNodesAsync()` on every dynamic group by query. Fixed in https://github.com/ravendb/ravendb/pull/15844/commits/62d63c3b99d86541210ace94fdc65416dc7cc3ad

arekpalinski committed:
RavenDB-18166 Introducing dedicated field to keep the raft index of auto index creatation operation so we won't execute WaitForExecutionOnAllNodesAsync() on every dynamic map-reduce query execution

ppekrol commented on src/Raven.Server/Json/BlittableJsonTextWriterExtensions.cs:
```diff
@@ -689,11 +689,11 @@ public static async Task<(long NumberOfResults, long TotalDocumentsSizeInBytes)>
             }
 
-            if (result.RaftCommandIndex.HasValue)
+            if (result.IndexDefinitionRaftIndex.HasValue)
```
@arekpalinski Shouldn't we re-add here the AutoIndexCreationRaftCommandIndex (and in the other place)?

arekpalinski replied:
Yes, added in https://github.com/ravendb/ravendb/pull/15844/commits/2dca5589dd9f322442496eee5b375c75cadc43fa

arekpalinski committed:
RavenDB-18166 We need to write AutoIndexCreationRaftCommandIndex to the response if it's not null and incorporate its maximum value from shards into combined sharded query result
